### PR TITLE
api/PostPolicy: Allow location header fully qualified URL

### LINKED ISSUE #4910: Allow full domain+path returned in Location header for single-endpoint configurations
When making a file create request against a minio server configuration with one endpoint

## Expected Behavior (as returned by AWS S3)
Location header returns something like this:
 `<Location>https://{configured-domain}/our-bucket/images/1505248844958-happy_turtle.png</Location>`

## Current Behavior
 `<Location>/our-bucket/images/1505248844958-happy_turtle.png</Location>`

## Context
Our current client doesn't support manipulating this, so we have to change our client to add add the domain to the beginning after processing the request.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
req.Host is used to construct the final object location.

## Motivation and Context
Fixes #4910

## How Has This Been Tested?
Manually

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by harshavardhana** [comment]
Adding tests for `getObjectLocation` function, will be adding them soon.

File: cmd/api-response.go [hunk 1]
```diff
@@ -275,9 +275,26 @@ func getLocation(r *http.Request) string {
...
+	if tls {
+		return httpsScheme
+	}
+	return httpScheme
+}
+
+// getObjectLocation gets the fully URI of an object if
+// server is bound to a specific IP or hostname. If not
+// the URI is relative.
+func getObjectLocation(host, bucket, object string) string {
```
Comment (balamurugana) [inline cmd/api-response.go]: We would need to use something like below
```go
u := URL{
Scheme: getURLScheme(),
Host: ...,
Path: ...,
}

return u.String()
```

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: cmd/api-response.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/api-response.go]: Not as default .. see the reasoning why it's based on `--address` alone. Otherwise we need it as relative..

Comment (harshavardhana) [reply cmd/api-response.go]: @balamurugana can you take look again?

**Comment by harshavardhana** [comment]
ping @donatello waiting on your review as well.

File: cmd/api-response.go [hunk 2]
```diff
@@ -275,9 +275,26 @@ func getLocation(r *http.Request) string {
...
-	return "/" + bucketName + "/" + key
+// returns "https" if the tls boolean is true, "http" otherwise.
+func getURLScheme(tls bool) string {
+	if tls {
+		return httpsScheme
+	}
+	return httpScheme
+}
+
+// getObjectLocation gets the fully URI of an object if
```
Comment (donatello) [inline cmd/api-response.go]: s/fully/fully qualified/

**Review by donatello — APPROVED** [review]
LGTM - however commit message needs to mention that the location header is related to post-policy-handler. Otherwise, some context appears to be missing.

**Comment by balamurugana** [comment]
As this PR fixes the issue partially i.e. its hard to provide usable fully qualified URL.  Discussion is needed.

File: cmd/bucket-handlers.go [hunk 3]
```diff
@@ -603,7 +603,7 @@ func (api objectAPIHandlers) PostPolicyBucketHandler(w http.ResponseWriter, r *h
 			Bucket:   objInfo.Bucket,
 			Key:      objInfo.Name,
 			ETag:     `"` + objInfo.ETag + `"`,
-			Location: getObjectLocation(objInfo.Bucket, objInfo.Name),
+			Location: getObjectLocation(globalMinioAddr, objInfo.Bucket, objInfo.Name),
```
Comment (krishnasrinivas) [inline cmd/bucket-handlers.go]: globalMinioAddr gets set to `:9000` if no --address is passed

Comment (harshavardhana) [reply cmd/bucket-handlers.go]: Correct.. this is waiting on a different architectural problem see above.

**Comment by harshavardhana** [comment]
The PR is now updated to use MINIO_ENDPOINT environment value after discussion @donatello @balamurugana  - please check.

File: cmd/server-main.go [hunk 4]
```diff
@@ -61,6 +61,9 @@ ENVIRONMENT VARIABLES:
   REGION:
      MINIO_REGION: To set custom region. By default it is "us-east-1".
 
+  ENDPOINT:
+     MINIO_ENDPOINT: To set location endpoint of the instance.
```
Comment (balamurugana) [inline cmd/server-main.go]: I think `MINIO_HOST_NAME` would be nice.  However we will discuss and take it forward

Comment (harshavardhana) [reply cmd/server-main.go]: The reason i didn't want it as HOST_NAME is because ENDPOINT encourages providing PORT as well @balamurugana

**Review by donatello — APPROVED** [review]

**Comment by harshavardhana** [comment]
@balamurugana we can remove blocked label from this PR since the PR addresses the agreed style we want to address this?

**Comment by balamurugana** [comment]
@harshavardhana @abperiasamy is looking into the environment variable naming. Rest of the code is good to go.

**Comment by harshavardhana** [comment]
> @harshavardhana @abperiasamy is looking into the environment variable naming. Rest of the code is good to go.

I changed to MINIO_ACCESS_ENDPOINT

**Comment by harshavardhana** [comment]
The PR is updated to use `r.Host` as the endpoint and leverage X-Forward-Proto as scheme hint, if not default to what is configured by the server.  Avoids envs and addresses all the concerns.

**Review by donatello — APPROVED** [review]

File: cmd/api-response.go [hunk 5]
```diff
@@ -275,9 +275,25 @@ func getLocation(r *http.Request) string {
...
+}
+
+// getObjectLocation gets the fully qualified URL of an object.
+func getObjectLocation(host, proto, bucket, object string) string {
+	if proto == "" {
+		proto = getURLScheme(globalIsSSL)
+	}
+	u := url.URL{
+		Host:   host,
+		Path:   slashSeparator + pathJoin(bucket, object),
```
Comment (balamurugana) [inline cmd/api-response.go]: I think we could use  `path.Join("/", bucket, object)`?

**Review by balamurugana — COMMENTED** [review]
Minor comment

**Review by balamurugana — APPROVED** [review]
LGTM

**Pull request merged**